OPHD/UI/Core/RadioButtonGroup.cpp

```diff
@@ -0,0 +1,68 @@
+#include "RadioButtonGroup.h"
+
+#include "../../Cache.h"
+#include "../../Constants.h"
+
+#include <NAS2D/Utility.h>
+#include <NAS2D/Renderer/Renderer.h>
+#include <NAS2D/MathUtils.h>
+
+#include <algorithm>
+
+
+using namespace NAS2D;
+
+
+RadioButtonGroup::RadioButtonGroup(std::vector<ButtonInfo> buttonInfos)
+{
+	mRadioButtons.reserve(buttonInfos.size());
+	for(auto &buttonInfo : buttonInfos)
+	{
+		NAS2D::Vector<int> offset = {0, 13};
+		offset.y = mRadioButtons.size() * offset.y;
+
+		auto &button = mRadioButtons.emplace_back(this, buttonInfo.name, buttonInfo.delegate);
+		button.visible(visible());
+		button.position(mRect.startPoint() + offset);
+	}
+}
+
+void RadioButtonGroup::onMove(NAS2D::Vector<int> displacement)
+{
+	Control::onMove(displacement);
+
+	for (auto &control : mRadioButtons)
+	{
+		control.position(control.position() + displacement);
+	}
+}
+
+void RadioButtonGroup::update()
+{
+	if (!visible()) { return; }
+
+	for (auto &control : mRadioButtons)
+	{
+		control.update();
+		control.hasFocus(true);
+	}
+}
+
+void RadioButtonGroup::clearSelection()
+{
+	for (auto &control : mRadioButtons)
+	{
+		control.checked(false);
+	}
+	mIndex = constants::NO_SELECTION;
+}
```

Comment:
This can probably be implemented without a loop. Assuming the at-most-one invariant is maintained, you could borrow the unselect code from the `setSelected` method. Actually, the `setSelected` method could delegate to `clearSelection`, so you only need the code in one place.

Reply:
There's a problem here, the `index` parameter is actually only used when building the control, it is not used when other interactions are done. Meaning there is no mapping from `int` to a valid RadioButton in the vector.

Naively you could pass a RadioButton `this` pointer to the RadioButtonGroup and have a lookup table of some sort, but that might be more complicated than needed.

Reply:
The `mIndex` field already tracks the currently selected index. Here I'm assuming it's always set correctly. Is there some doubt about this? The code in `setSelected` seems to assume it is set correct.

Code to borrow from `setSelected` to implement `clearSelection`, with the last line updated:
```cpp
if (mIndex != constants::NO_SELECTION)
{
    mRadioButtons[mIndex].checked(false);
}
mIndex = constants::NO_SELECTION;
```

----

I suppose this relates to why I expected there to be two separate callbacks when a button is selected. One for the button specific interaction, and another for the `RadioButtonGroup` to ensure at-most-one-hot. This also relates to why I expected `clearSelection` to be called from an event handler on the `RadioButtonGroup`, rather than through the button with `mParentContainer->clearSelection();`.

If you remember, I had expected the `mRbgDelegate()` to call the at-most-one-hot event handler on the `RadioButtonGroup`:
```cpp
mParentContainer->clearSelection();
mRbgDelegate();
```

If that had been the case, it would have made sense to call `clearSelection()` from within the event handler on `RadioButtonGroup`, rather than just before the event handler was called. Though it seems `mRbgDelegate` really ended up being `mSignal`, which was a different concept entirely.

----

Anyway, my thought is, when a new button is selected the `RadioButtonGroup` somehow knows this, probably by calling an event handler method. When you clear a selection, the `RadioButtonGroup` will know what the previously selected button was, and can unselect it.

Looking back at the code now, it seems there is currently no code to inform the `RadioButtonGroup` of the update. Maybe we do need something like an `mRgbOneHot` delegate (similar to what I thought `mRgbDelegate` was supposed to be).

As you've mentioned, we can't just call `mParentContainer->setSelected` since the child control doesn't know it's own index. We could substitute with the `RadioButton` control address (`this`). Since the `RadioButton` controls are all in the same `std::vector`, it's is possible to convert an element address to an index in constant time (no loops).

Reference:
https://www.geeksforgeeks.org/computing-index-using-pointers-returned-stl-functions-c/

Reply:
Actually, I was just thinking. We don't really need a fully general `Delegate` system for the at-most-one-hot method. We know for sure the target control is `RadioButtonGroup`, and what the exact method is, and the `RadioButton` control is already tied to `RadioButtonGroup` in terms of design. Since it already has a `mParentContainer` reference, it can just call the event handler directly. Could potentially even make the event handler `private` or `protected`, and give `friend` access to the `RadioButton` control. That would lock down access to the event handler, and prevent abuse from outside code.

Or perhaps instead of an event handler style method, we could have an overload of `setSelected`, which takes a `RadioButton` control address instead of an index.

```cpp
mParentContainer->setSelected(this);
```

The method could use the technique from the article linked above to convert the address to an index (basically subtract the control address from the array base address), and then delegate to the index overload of `setSelected`.

Reply:
Hi @DanRStevens I checked out the link provided, but that seems to detail the std::max_elements function, whereas we probably need something like this which can find a "random" entry in a vector based on its element pointer, https://thispointer.com/c-how-to-find-an-element-in-vector-and-get-its-index/ using std::find and std::distance (but I assume the find loops over the container). Would the std::find and std::distance combination be okay to use here to go from the pointer of the RadioButton to its index?

Reply:
I just doubled checked the link I sent you earlier. The article makes use of `std::max_element` in their example, but that's just an example problem where you would get an element address, rather than an element index. The bulk of the article is about converting that element address into an element index.

In the example they use the expression:
```cpp
max_element(vi.begin(), vi.end()) - vi.begin()
```

Though really the essential point is (for an element within an array):
```cpp
auto index = elementAddress - arrayAddress; // There is an implicit division by the element size
```

Remember that if you use `p++` on a pointer, it increments the address by the size of the pointed to element, so internally it would compute `p + 1 * sizeof(*p)`. Similarly if you were to write `p + 2` for a pointer, it really means `p + 2 * sizeof(*p)`. When you subtract pointers, it does the reverse operation, including the divide, so `(p + 2) - p == 2`.

Using [`std::distance`](https://en.cppreference.com/w/cpp/iterator/distance) is basically the same as subtracting pointers. Or at least, for *LegacyRandomAccessIterator* it's the same as subtracting pointers. Think contiguous containers, such as `std::array` and `std::vector`, as opposed to non-contiguous containers, such as `std::list`, or any of the tree data structures. For more general pointer types, `std::distance` can abstract the same concept of element distance, where the result is not actually subtracting pointers, but may require iterator traversal, such as for a tree data structure. In the case of a tree, the memory for the various nodes may not be at well defined address offsets from each other, especially after a tree rebalancing operation, though there is still a defined concept of distance, which can be determined by traversing the tree.

----

The article you linked also deals with converting an element address to an index. Though in that article they use an example with `std::find`, which returns an element address, which they then want to convert to an index. The use of `std::find` isn't relevant to the conversion from address to index. It's just an example function that will give you an address, which you might want to convert to an index. The particulars of `std::find` are irrelevant to the conversion, much like the result of `max_element` was. It's just that they both returned addresses rather than indexes. It's `std::distance` that does the actual conversion.

----

Btw, fun fact, if you want to divide, and you know there will be no remainder, then there is a simplification when working modulo a register size. Instead of dividing, you can find a multiplicative inverse (modulo the register size), and a shift value, and instead multiple (with wrap around) and shift to get the effect of division. That means you can divide by using the much faster multiply and shift instructions, rather than the slow and long delayed division instruction. This is particularly useful with arrays since two element addresses within the same array will always differ by a multiple of the element size, and so if you subtract them and divide, there will never be any remainder. Hence the multiplicative inverse technique will apply. Of course, if the division is a power of two, you only need the shift.

There was a section in Hacker's Delight about using multiplicative inverses to do division.

Reply:
Hi @DanRStevens Thanks for that and you were completely right! I didn't know that was possible! Another thing learned!
I'll also keep in mind the multiplicative inverses, that's a nice win.

I've now implemented it as such:

```
void setSelected(std::size_t index);
void setSelected(RadioButtonGroup::RadioButton* button);
```

```
void RadioButtonGroup::setSelected(std::size_t index)
{
	clearSelection();
	mIndex = index;
	mRadioButtons[index].checked(true);
}
```

```
void RadioButtonGroup::setSelected(RadioButtonGroup::RadioButton* button)
{
	auto index = button - mRadioButtons.data();
	setSelected(index);
}
```